[develop] Add RRFS product generation job to the SRW app

BenjaminBlake-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
This pull request adds the RRFS product generation (prdgen) job to the SRW app.  The prdgen job uses wgrib2 to generate various subdomain RRFS grib2 files from the raw-UPP generated grib2 output from the run_post task.  It will be needed for the real-time operational RRFS runs, which will be over the RRFS_NA_3km grid.  In the config.yaml file, two new configuration flags have been added.  RUN_TASK_RUN_PRDGEN needs to be set to true if you want to run the prdgen job, and DO_PARALLEL_PRDGEN can be set to true if you are running over the RRFS_NA_3km grid and want to use CFP to speed up the runtime of this task.  Two ex-scripts have also been added - exregional_run_prdgen.sh, which is the main script, and exregional_run_prdgen_subpiece.sh, which is only called if DO_PARALLEL_PRDGEN = TRUE.  In addition, a new WE2E test was added to the wflow_features directory for running the prdgen job over the RRFS_NA_3km grid.

A couple of notes about this task:
1. A testbed text file is optionally needed for this job to create testbed files.  In the RRFS workflow, it is included in sorc/UPP/parm, which gets linked into parm/upp.
2. RUN_CMD_PRDGEN has been added to all the machine.yaml files, but may need to be modified in the future for all machines besides WCOSS2.  RUN_CMD_PRDGEN is invoked when DO_PARALLEL_PRDGEN = TRUE and therefore will need to include a call to cfp.
3. If DO_PARALLEL_PRDGEN = TRUE with the RRFS_NA_3km grid, then PPN_RUN_PRDGEN needs to be 22 and not 1.

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X ] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [X ] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [X ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [X ] comprehensive tests (grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16, grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR, MET_verification, community_ensemble_2mems_stoch, prdgen)

## DEPENDENCIES:
N/A

## DOCUMENTATION:
The documentation will need to be updated but I don't have any experience with doing this myself.

## ISSUE: 
Fixes issue [#580 ](https://github.com/ufs-community/ufs-srweather-app/issues/580)

## CHECKLIST
<!-- Add an X to check off a box. -->
- [X ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [X ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [X ] My changes generate no new warnings
- [X ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 
<!-- If you do not have permissions to add labels to your own PR, request that labels be added here. 
Add an X to check off a box. Delete any unnecessary labels. -->
A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [X ] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [X ] help wanted

## CONTRIBUTORS (optional): 
@hu5970, @ShunLiu-NOAA, @danielabdi-noaa

BenjaminBlake-NOAA committed:
Add prdgen J-job

BenjaminBlake-NOAA committed:
Initial updates from Daniel's feature/process_bufr branch

BenjaminBlake-NOAA committed:
Add run_prdgen configure option and add task to xml file

BenjaminBlake-NOAA committed:
Updates to prdgen scripts and config files

BenjaminBlake-NOAA committed:
Remove mem_run_prdgen from FV3LAM_wflow.xml

BenjaminBlake-NOAA committed:
Adding WE2E test for prdgen job

Commit:
Fix resource settings for prdgen job

Commit:
Updating the machine.yaml files

christinaholtNOAA commented on jobs/JREGIONAL_RUN_PRDGEN:
```diff
@@ -0,0 +1,140 @@
+if [ ${CYCLE_TYPE} == "spinup" ]; then
+  DATA="${DATA:-${COMIN}${SLASH_ENSMEM_SUBDIR}}"
+else
+  DATA="${DATA:-${COMIN}${SLASH_ENSMEM_SUBDIR}}"
```
Should these two be the same? If so, then can the if statement be removed?

MichaelLueken replied:
It looks like line 67 be:

DATA="${DATA:-${COMIN}${SLASH_ENSMEM_SUBDIR}}/fcst_fv3lam_spinup"

while line 69 should be:

DATA="${DATA:-${COMIN}${SLASH_ENSMEM_SUBDIR}}/fcst_fv3lam".

BenjaminBlake-NOAA replied:
I think eventually that will be the plan once more RRFS tasks/features are added, but for now I'll remove this if statement.

christinaholtNOAA commented on parm/FV3LAM_wflow.xml:
```diff
@@ -753,6 +760,49 @@ the <task> tag to be identical to the ones above for other output times.
+      <envar><name>SLASH_ENSMEM_SUBDIR</name><value><cyclestr>{{ slash_ensmem_subdir }}</cyclestr></value></envar>
+      <envar><name>ENSMEM_INDX</name><value><cyclestr>#{{ ensmem_indx_name }}#</cyclestr></value></envar>
+
+      <envar><name>CYCLE_TYPE</name><value><cyclestr>prod</cyclestr></value></envar>
```
Would this be set differently, or is it always `CYCLE_TYPE=prod`? If it is more or less always the same, does it make sense to move it to be hard-coded in the ex-script? Is it an NCO requirement to be exported?

I suppose the same questions hold for `CYCLE_SUBTYPE` and `TMMARK` below.

BenjaminBlake-NOAA replied:
These variables can be revisited in a later PR, I'll remove them for this one.

christinaholtNOAA commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,437 @@
+# 2 subpieces for Hawaii and Puerto Rico grids
+sed -n -e '1,500p' $DATAprdgen/prslevf${fhr}.txt >& $DATAprdgen/hi_pr_1.txt
+sed -n -e '501,$p' $DATAprdgen/prslevf${fhr}.txt >& $DATAprdgen/hi_pr_2.txt
+
```
Would a loop over the following section work to reduce repetition of information? I was thinking something like (warning! totally untested bash):
```
echo "#!/bin/bash" > $DATAprdgen/poescript_${fhr}
echo "export DATA=${DATAprdgen}" >> $DATAprdgen/poescript_${fhr}
echo "export COMOUT=${COMOUT}" >> $DATAprdgen/poescript_${fhr}

tasks=(4 4 2 2 10)
domains=(conus ak hi pr namerica)
for i in $(seq ${#domains[@]}), do
  domain=${domains[$i]}
  for task in $(seq ${tasks[$i]}); do 
    mkdir -p $DATAprdgen/prdgen_${domain}_${task}
    echo "$SCRIPTSdir/exregional_run_prdgen_subpiece.sh $fhr $cyc $task $domain ${DATAprdgen} ${COMOUT} &" >> $DATAprdgen/poescript_${fhr}
    
  done
done
echo "wait" >> $DATAprdgen/poescript_${fhr}
chmod 775 $DATAprdgen/poescript_${fhr}
```

BenjaminBlake-NOAA replied:
Yes, I think something like this should work!  Good idea.  I can test it next week after I return from leave.

BenjaminBlake-NOAA replied:
Just committed a revised version of the script.

christinaholtNOAA commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,437 @@
+# reassemble the output grids
+#----------------------------------------
+
+cat $DATAprdgen/prdgen_conus_1/conus_1.grib2 \
```
A similar loop could be applied in this section, too.

BenjaminBlake-NOAA replied:
Agreed, will look into doing that.

christinaholtNOAA commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,437 @@
+#
+# background files
+#
+prslev=${COMOUT}/${NET}.${cycle}${dot_ensmem}.prslev.f${fhr}.${POST_OUTPUT_DOMAIN_NAME}.grib2
```
Where does `dot_ensmem` get set? Apologies if I missed that one.

MichaelLueken replied:
@christinaholtNOAA It looks like `dot_ensmem` is set in `ush/job_preamble.sh` in the current `develop` branch:

https://github.com/ufs-community/ufs-srweather-app/blob/b34ab05a13d1fc571afd5f99f790e5ec71bc9a14/ush/job_preamble.sh#L27-L31

christinaholtNOAA commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,437 @@
+      mkdir -p ${COMOUT}/${grid}_grid
+      cp_vrfy ${bg_remap} ${COMOUT}/${grid}_grid/${NET}.${cycle}${dot_ensmem}.${leveltype}.f${fhr}.${POST_OUTPUT_DOMAIN_NAME}.grib2
+
+      if [ $leveltype = 'prslev' ]; then
```
I may have overlooked it, but each of these if statements contain roughly the same information. Would this one command work to replace them all? I've replaced the hardcoded `prslev` with `${leveltype}` instead.

```
         ln_vrfy -fs --relative ${COMOUT}/${grid}_grid/${NET}.${cycle}${dot_ensmem}.${leveltype}.f${fhr}.${POST_OUTPUT_DOMAIN_NAME}.grib2 ${COMOUT}/${net4}.${cycle}.${leveltype}.f${fhr}.${gridname}grib2
         wgrib2 ${COMOUT}/${net4}.${cycle}.prslev.f${fhr}.${gridname}grib2 -s > ${COMOUT}/${net4}.${cycle}.${leveltype}.f${fhr}.${gridname}grib2.idx
```

BenjaminBlake-NOAA replied:
Yes I agree.  I'm about to test a new version of the script with this change and the ones you recommended above with adding loops.  Once that's done I'll commit the changes and mark these comments as resolved.

christinaholtNOAA commented on scripts/exregional_run_prdgen_subpiece.sh:
```diff
@@ -0,0 +1,63 @@
+  -new_grid_interpolation neighbor \
+  -if ":(WEASD|APCP|NCPCP|ACPCP|SNOD):" -new_grid_interpolation budget -fi \
+  -new_grid ${gridspecs} ${domain}_${subpiece}.grib2
+
```
I think the lines below this point may be unnecessary.

BenjaminBlake-NOAA replied:
I agree, I will remove them.

christinaholtNOAA commented on tests/WE2E/test_configs/wflow_features/config.prdgen.yaml:
```diff
@@ -0,0 +1,43 @@
+  RUN_ENVIR: community
+workflow:
+  CCPP_PHYS_SUITE: FV3_RRFS_v1beta
+  PREDEF_GRID_NAME: RRFS_NA_3km
```
That's an expensive test! Could we leverage any of the existing tests (if we already have a NA domain test) to also test this task? Could we alternatively test just the prodgen task with staged data?

BenjaminBlake-NOAA replied:
There is one WE2E test for the RRFS_NA_3km grid in the grids_extrn_mdls_suites_community directory.  It is called grid_RRFS_NA_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta, so it is very similar to the one I added here.  I'm willing to add the prdgen job to this pre-existing WE2E test if that makes sense and I can remove this one.  What do you think?

BenjaminBlake-NOAA replied:
I added the prdgen job to the pre-existing WE2E task for the RRFS_NA_3km grid.

RatkoVasic-NOAA commented on scripts/exregional_run_prdgen_subpiece.sh:
```diff
@@ -0,0 +1,63 @@
+  -if ":(WEASD|APCP|NCPCP|ACPCP|SNOD):" -new_grid_interpolation budget -fi \
+  -new_grid ${gridspecs} ${domain}_${subpiece}.grib2
+
+#export err=$?; err_chk
```
Is this supposed to be commented out?

BenjaminBlake-NOAA replied:
Yes it is not needed, I will remove the line.

christinaholtNOAA commented on ush/machine/odin.yaml:
```diff
@@ -9,8 +9,11 @@ platform:
   PARTITION_HPSS: workq
   QUEUE_HPSS: workq
+  PARTITION_PRDGEN: workq
+  QUEUE_PRDGEN: workq
```
Would it work for the prdgen task to use `QUEUE_` and `PARTITION_DEFAULT` instead of introducing a new queue? Are there platforms that may need a totally separate queue to do the thing for this job that's different than others?

I think we've accumulated 3 so far for a catchall (DEFAULT), the forecast (because we may want beefier nodes than a default on the cloud, or on Jet), and then HPSS for access nodes that have connectivity to the internet.

BenjaminBlake-NOAA replied:
Yes I think it should.  I will use QUEUE_DEFAULT and PARTITION_DEFAULT for this task.

christinaholtNOAA reviewed (commented):
@BenjaminBlake-NOAA This is great! Thanks for preparing this. I have just a few suggestions and questions below.

chan-hoo commented:
@BenjaminBlake-NOAA, 1) Please merge the `develop` branch into your branch. There are some conflicts with the latest develop branch. 2) What do you think about changing the task name from `run_prdgen` to `gen_product` or similar?

ShunLiu-NOAA commented:
@BenjaminBlake-NOAA and @chan-hoo, Good suggestion about the task name. It is reasonable to remove "run_" in job name. Maybe we can just use "prdgen". We also consider removing "run_" for "prestart", "post", "fcst", "enkf" and "analysis" job in regional workflow in RRFS_DEV1.

BenjaminBlake-NOAA committed:
Merging changes from latest develop branch

BenjaminBlake-NOAA commented:
@chan-hoo  I just merged all the changes from the develop branch into my feature branch, hopefully I did that right.  I agree with @ShunLiu-NOAA and think we could call this task the "prdgen" task instead of "run_prdgen".  The operational NAM has a prdgen task.  @MatthewPyle-NOAA  what are your thoughts about the name of this task?

@christinaholtNOAA and @RatkoVasic-NOAA , thank you for your suggestions and feedback!  They all sound quite reasonable to me and I will try and address as many of them as I can before I go on leave (tomorrow through next Tuesday).

JeffBeck-NOAA commented:
> @BenjaminBlake-NOAA and @chan-hoo, Good suggestion about the task name. It is reasonable to remove "run_" in job name. Maybe we can just use "prdgen". We also consider removing "run_" for "prestart", "post", "fcst", "enkf" and "analysis" job in regional workflow in RRFS_DEV1.

@gsketefian, thoughts on this?

BenjaminBlake-NOAA committed:
Fixing issue in config_defaults.yaml

BenjaminBlake-NOAA committed:
Remove QUEUE_PRDGEN and PARTITION_PRDGEN, and delete unneeded lines from ex-script

gsketefian commented:
@JeffBeck-NOAA I like having task names with verbs in them, like "run_this_or_that" or "gen_something" (generate) because it allows for a more descriptive/informative name.  But I suppose they can all be nouns, like "fcst" instead of "run_fcst", as long as the naming convention is consistent (verb-like or noun-like).  But I suspect this path will lead to very short names that sound cryptic to a community user of the SRW who's not involved in the day-to-day weeds of development.  In this particular case, for now I'd go with `gen_product` as @chan-hoo suggests to be consistent with how the other tasks are named (verb-like).  If task renaming is something people are interested in discussing, we should do so at the next code management meeting and/or open an issue in the repo.

JeffBeck-NOAA commented:
Thanks, @gsketefian!  I agree that we should keep it consistent across task names and maintain a descriptive quality for those who aren't as familiar with the workflow.

chan-hoo commented:
@JeffBeck-NOAA, @gsketefian, I basically agree with you about the consistent and readable naming. However, the use of the prefix `run_` is questionable to me. The task switch names in `config_default.yaml` start with `run_task_`. If the task name begins with `run_`, the switch name becomes `run_task_run_`. This looks a little bit weird to me. We know that the tasks will run without the prefix `run_`. This is the reason why all the extra tasks for air quality modeling (Online-CMAQ) do not have a verb on them. I agree that `prdgen`, `enkf`, `analysis`, and even `fcst` look too short.

gsketefian commented:
@chan-hoo Yes, I agree `run_task_run_fcst` is a bit strange.  The first `run_` can be changed to something else, like `incl_` for "include" or `act_` or `actv_` for "activate" a task in the workflow, so those flags would be renamed `incl_task_run_fcst` or `actv_task_run_fcst`.

chan-hoo commented:
@gsketefian, @JeffBeck-NOAA, what do you think about the consistency to 'noun-like' naming? If we put the prefix `run_` to the scripts, the majority of the tasks for RRFS and AQM will have it. I think it will make the task name too long. Many other applications do not use a verb on the file names of their scripts. For example, you can check the global workflow (https://github.com/NOAA-EMC/global-workflow/tree/develop/scripts). Even in the example of the j-job or ex-script in the NCO standards, the file name is `JPMB_FORECAST` or `expmb_forecast.sh` (I don't think it is a requirement though). In my opinion, it would be good to have consistency with other applications. @MatthewPyle-NOAA, @aerorahul, can you let us know how other applications name the j-job and ex- script files? I'd like to hear from others too; @MichaelLueken, @mkavulich, @christinaholtNOAA, @danielabdi-noaa, @ShunLiu-NOAA, @hu5970.

MatthewPyle-NOAA commented:
@chan-hoo I'd say there isn't a single standard, but most follow something close to this idea.  For a production model system we'll call "ABC":

J-job file name would start with a `JABC_` string
ex-script file names would have a format like `exabc_*.sh`

There will be a correspondence between the J-jobs and ex-scripts, such as:

JABC_POST would execute something like exabc_post.sh

Personally I think the inclusion of "run" in task names is superfluous.

JeffBeck-NOAA commented:
@MatthewPyle-NOAA, @chan-hoo, these are fair points.  I think "noun-like" naming would be fine.  The only comment I'd add is that, not being familiar with the global workflow, when I see "JPMB_FORECAST", I don't know what "pmb" stands for.  So, if possible, it'd be good to have a more descriptive name, without making the name too long, as @chan-hoo mentioned.  Our current configuration has "JREGIONAL" and "exregional" for the j-jobs and ex-scripts prefixes, which I think follows the general standard.

aerorahul commented:
> @chan-hoo I'd say there isn't a single standard, but most follow something close to this idea. For a production model system we'll call "ABC":
> 
> J-job file name would start with a `JABC_` string ex-script file names would have a format like `exabc_*.sh`
> 
> There will be a correspondence between the J-jobs and ex-scripts, such as:
> 
> JABC_POST would execute something like exabc_post.sh
> 
> Personally I think the inclusion of "run" in task names is superfluous.

The convention is `JRUN_`  where `RUN` is the "name of the model run (third level of the COM directory structure).
If a j-job serves multiple values for "RUN", then something common e.g. `JGLOBAL_`, instead of `JGDAS_` or `JGFS_` is used.

Each `JRUN_` has a corresponding `exRUN_`.  Same logic applies when an ex-script serves multiple values for "RUN"

The top-level job card `ecf` driver of the `JRUN_` scripts is named uniquely e.g. `jRUN_.ecf`  There is a unique job card `ecf` driver for each value of `RUN`.

Please see Section VII of the EE2 standards.
"All examples are for job jpmb_forecast. Model name is nco and type of model run is pmb"
This statement implies `RUN = pmb`

gsketefian commented:
@chan-hoo If you're proposing changing to noun-like task name, can you create an issue (like the one I did for renaming of vx tasks, #619) so we can see exactly what the new names will be of each task?  There are already issues that come to my mind.  For example, if `get_extrn_ics` becomes `extrn_ics` and `make_ics` becomes `ics`, it will not be clear (especially to a relatively inexperienced user) what the difference between these two is.  I suspect that after considering all the tasks, it's going to be difficult to stick to a strictly noun-like or verb-like naming convention AND maintain clarity AND have names that aren't too long, so it may end up being a mix of nouns and verbs.  And I strongly suggest doing this in a separate PR, since it will be a significant change.

chan-hoo commented:
@gsketefian, sure, I'll open an issue for this soon.

Commit:
Removing CYCLE_TYPE, this can be added in a later PR when RRFS spinup tasks are added

Commit:
Remove variables from FV3LAM_wflow.xml

Commit:
Suggested script changes from Christina and other bug fixes

BenjaminBlake-NOAA committed:
Merge branch 'develop' into feature/prdgen

Commit:
Resolving conflicts with xml in develop branch

Commit:
Adding prdgen job to pre-existing WE2E test

Commit:
Remove prdgen WE2E test

christinaholtNOAA commented on jobs/JREGIONAL_RUN_PRDGEN:
```diff
@@ -0,0 +1,127 @@
+  export COMOUT="${COMOUT}"
+else
+  export COMOUT="${DATA}/postprd"
+fi
```
This could be simplified to:

```
if [ "${RUN_ENVIR}" = "community" ]; then
  export COMOUT="${DATA}/postprd"
fi
mkdir_vrfy -p "${COMOUT}"
```

BenjaminBlake-NOAA replied:
Agreed, making this change now.

christinaholtNOAA commented on parm/FV3LAM_wflow.xml:
```diff
@@ -1188,6 +1189,46 @@ the <task> tag to be identical to the ones above for other output times.
   {%- endif %}
 {%- endif %}
 {%- endif %}
+{%- if run_task_run_prdgen %}                      
```
Just checking...is this only meant to run for the hourly graphics (as opposed to subhourly)? _Yes_ is totally fine, just curious.

BenjaminBlake-NOAA replied:
Eventually we will likely want to run the prdgen job for subhourly output too, but initially with this PR I think it is fine to only include it for hourly output.

christinaholtNOAA commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,366 @@
+tasks=(4 4 2 2 10)
+domains=(conus ak hi pr namerica)
+count=0
+for i in ${domains[@]}
```
It's a little more readable with 

```
for domain in ${domains[@]}
```

BenjaminBlake-NOAA replied:
Yes that looks better, I'll modify both loops to use for domain in ${domains[@]}

christinaholtNOAA commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,366 @@
+
+DATA=$COMOUT
+DATAprdgen=$DATA/prdgen_${fhr}
+mkdir $DATAprdgen
```
There are several instances of `mkdir` instead of `mkdir_vrfy` that is more commonly used.

BenjaminBlake-NOAA replied:
I'll update those to mkdir_vrfy.

christinaholtNOAA commented on ush/config_defaults.yaml:
```diff
@@ -1997,6 +2007,10 @@ task_run_post:
   #
+  # TESTBED_FIELDS_FN
+  # The file which lists grib2 fields to be extracted to bgsfc for testbed
+  # Empty string means no need to generate bgsfc for testbed
```
Would you mind clarifying what bgsfc means here in the comment?

BenjaminBlake-NOAA replied:
bgsfc is an old naming convention that will no longer be used going forward.  I'll change this to something like: "The file which lists grib2 fields to be extracted for testbed files.  Empty string means no need to generate testbed files."

christinaholtNOAA reviewed (approved):
@BenjaminBlake-NOAA This looks really good. I left a couple of nit-picky comments/suggestions below, but overall, I think it looks great. Thanks for addressing my previous comments.

Commit:
Update documentation in ConfigWorkflow.rst

Commit:
Addressing Christina's suggested changes to J-job and ex-script

BenjaminBlake-NOAA commented:
I've addressed all of @christinaholtNOAA 's suggestions and added documentation for the run_prdgen variables to ConfigWorkflow.rst with @gspetro-NOAA 's help.  The Python unittests and functional tests have passed, so I think this PR is now ready to be merged.  Please review/approve when you have time - thanks!

christinaholtNOAA reviewed (approved):
LGTM! Thanks @BenjaminBlake-NOAA

danielabdi-noaa reviewed (dismissed):
I believe this should be close to the last PR to go in so that it can be tested it works well the `run_post` job and the `run_pygraf` job that comes after it. We can't be sure it works well with them unless it is tested that way. That is a major downside of the standalone approach. For example:
  - The changes I made to make "spinup" and "prod" cycles is removed from ex- and j-job scripts. That guarantees it will not work well with `run_post` and `run_pygraf`
  - The workflow added with this PR is rudimentary to say the least. It just gets in the way later IMO.
  - The "standalone" testing approach is not demonstrated with this PR. Where are the inputs, where are the standalone scripts to test it with?

gspetro-NOAA commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -1222,6 +1225,48 @@ Set parameters for customizing the :term:`UPP`.
+   Controls the size of the stack for threads created by the OpenMP implementation.
+
+``DO_PARALLEL_PRDGEN``: (Default: false)
+   Flag that determines whether to use CFP to run the product generation job in parallel.  This should be used with the RRFS_NA_3km grid and PPN_RUN_PRDGEN should be set to 22.
```
@BenjaminBlake-NOAA What is CFP? Better to define for newbies even if seasoned forecasters would know. 
Also, could you add backticks around RRFS_NA_3km grid and PPN_RUN_PRDGEN?
Thanks! :)

BenjaminBlake-NOAA replied:
Sure I can add backticks to those two variables.  CFP allows you to run multiple executions of code at the same time, but I can't seem to find what it actually stands for!  Will attempt to track that down...

BenjaminBlake-NOAA replied:
I couldn't actually find what the CFP acronym stands for but I'll add a sentence describing what CFP is used for.

gspetro-NOAA commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -1222,6 +1225,48 @@ Set parameters for customizing the :term:`UPP`.
+---------------------------------
+For each workflow task, certain parameter values must be passed to the job scheduler (e.g., Slurm), which submits a job for the task.
+
+``TN_RUN_PRDGEN``: (Default: "run_prdgen")   Set the name of this Rocoto workflow task. Users typically do not need to change this value.
```
@BenjaminBlake-NOAA I think you might need a new line before the definition on this one.

BenjaminBlake-NOAA replied:
Thanks for catching that, I'll get it fixed.

gspetro-NOAA commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -1222,6 +1225,48 @@ Set parameters for customizing the :term:`UPP`.
+``ADDNL_OUTPUT_GRIDS``: (Default: [])
+   Set additional output grids for wgrib2 remapping, if any.  Space-separated list of strings, e.g., ( "130" "242" "clue").  Default is no additional grids.
+
+``TESTBED_FIELDS_FN``: (Default: "")
```
@BenjaminBlake-NOAA I think config_defaults.yaml had a TESTBED_FIELDS_FN2 field as well (but it looks like that might be gone now?), and both TESTBED_FIELDS_FN* were defined in the MAKE_GRID Config Params section. Is there a reason the second one has been omitted and the first moved to this section?

BenjaminBlake-NOAA replied:
I think what happened was I realized that defining them in the MAKE_GRID Config Params section was a mistake, and since it is only needed for the RUN_PRDGEN task I moved it to there.  Originally there was also a TESTBED_FIELDS_FN2 variable for defining what parameters should go into the testbed files, but I think initially it makes sense to just add one TESTBED_FIELDS_FN variable at this time.

Commit:
Updates to ConfigWorkflow.rst

BenjaminBlake-NOAA commented:
@danielabdi-noaa  I am fine with waiting to merge this PR until later, but I was under the impression that we were interested in merging it soon.  I removed the changes related to the "spinup" and "prod" cycles because those tasks haven't been added to the SRW app yet, so it didn't make sense to include them in this PR.  @christinaholtNOAA do you think it makes sense to merge this work now or wait until other RRFS tasks have been added?

christinaholtNOAA commented:
@danielabdi-noaa this task can be added to the existing workflow cleanly, doesn't depend on any other additions, and has been demonstrated here with a WE2E test addition. It takes in the output of an existing task and creates products. It is a rudimentary task, so that it's implementation is simple makes sense.

When added to the existing workflow, a new task doesn't necessarily need a standalone method of testing. I've been suggesting that some tasks use standalone testing to avoid a huge expansion of logic associated with the XML generation. This one does not cause any of those issues.

The addition of `spinup` and `prod` cycles for RRFS will essentially be at the workflow level, and I don't think that having prod gen enabled should not change any of that. The dependencies and cycles may need to be changed, just like any other task that goes into the workflow when we are defining those cycles.

@BenjaminBlake-NOAA I think it's a good idea to go ahead with the merge now before your branch goes stale, causing you additional work.

MichaelLueken commented:
Since this is the next component that will be merged into the SRW App for the RRFS implementation, I would like to ask @gspetro-NOAA and @RatkoVasic-NOAA to review the changes one last time and, if @BenjaminBlake-NOAA has made the desired changes, to please go ahead and approve the PR.

@danielabdi-noaa If @christinaholtNOAA's above explanation is okay, please also provide an approval once you are okay with the changes.

Thanks!

MichaelLueken reviewed (approved):
@BenjaminBlake-NOAA These changes look good to me!  I do have one quick question - I see that you added a `run_prdgen.local.lua` file for WCOSS2.  Do the other machines require a run_prdgen task modulefile as well, or is this only an issue for WCOSS2?  Will go ahead and approve now.

danielabdi-noaa commented:
@christinaholtNOAA  I disagree. If I had not tested this within a workflow, it would be a stretch to say this will work within the RRFS dev workflow. 

- I do not think a PR that removes the "spinup" and "prod" cycles in the ex-scripts and j-jobs, which will definately not work later, is a good idea.

- I also do not think adding bits of the workflow that will definately rewritten for "spinup" and "prod" cycles later is a good idea either.

These two just add work for the next PRs to come. It would be best to leave out the changes to the xml file and add back the support for "spinup" and "prod" cycles in the scripts.

Also it is good to know that the strictly "standalone" approach is not a must have.

Having raised my concerns, I will remove my request for change though I will not approve it.

BenjaminBlake-NOAA commented:
@MichaelLueken  Thanks for reviewing the changes!  The other machines may also require a run_prdgen.local.lua file, unless we only plan on supporting the RRFS_NA_3km grid on WCOSS2.  That file is needed to load the CFP module when using DO_PARALLEL_PRDGEN = true with the RRFS_NA_3km grid.

danielabdi-noaa commented on tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_NA_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta.yaml:
```diff
@@ -37,3 +37,7 @@ task_run_fcst:
   NNODES_RUN_POST: 8
   PPN_RUN_POST: 12
+task_run_prdgen:
+  RUN_TASK_RUN_PRDGEN: true
```
This should be in `workflow_sections`.

christinaholtNOAA replied:
I believe that is `workflow_switches`

BenjaminBlake-NOAA replied:
Yes it looks like it should fall under workflow_switches.  I'll fix that along with your other suggested changes below and push the fixes to GitHub soon - thanks for reviewing!

danielabdi-noaa commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,364 @@
+for leveltype in prslev natlev ififip testbed
+do
+  if [ -f ${COMOUT}/${NET}.${cycle}${dot_ensmem}.${leveltype}.f${fhr}.${POST_OUTPUT_DOMAIN_NAME}.grib2 ]; then
+    ln_vrfy -sf --relative ${COMOUT}/${NET}.${cycle}${dot_ensmem}.${leveltype}.f${fhr}.${POST_OUTPUT_DOMAIN_NAME}.grib2 ${COMOUT}/${net4}.${cycle}.${leveltype}.f${fhr}.${gridname}grib2
```
There was a question whether to use `POST_OUTPUT_DOMAIN_NAME` or `TMMARK`. Last time I checked with Ming Hu, he mentioned we should use the latter even though I thought that `POST_OUTPUT_DOMAIN_NAME` fits with SRW's way of doing things better. May want to check with him again.

BenjaminBlake-NOAA replied:
@hu5970 Are you ok with us using POST_OUTPUT_DOMAIN_NAME or would TMMARK be preferred?  I think I agree with @danielabdi-noaa and would rather use POST_OUTPUT_DOMAIN_NAME.

danielabdi-noaa commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,364 @@
+net4=$(echo ${NET:0:4} | tr '[:upper:]' '[:lower:]')
+for leveltype in prslev natlev ififip testbed
+do
+  if [ -f ${COMOUT}/${NET}.${cycle}${dot_ensmem}.${leveltype}.f${fhr}.${POST_OUTPUT_DOMAIN_NAME}.grib2 ]; then
```
I have not tested the naming of files with "ensemble" members properly -- because RRFS CONUS 3km did not have it IIRC. So the `${dot_ensmem}` additions are not tested although I have tested the change to `f${fhr}` from `${fhr}` which was represented as `{fhr}` works

danielabdi-noaa commented on scripts/exregional_run_prdgen.sh:
```diff
@@ -0,0 +1,364 @@
+    echo "${PARMdir}/upp/${TESTBED_FIELDS_FN} not found"
+  fi
+fi
+if [[ ! -z ${TESTBED_FIELDS_FN2} ]]; then
```
This is a bug if `TESTBED_FIELDS_FN2` is not defined? I think @gspetro-NOAA also poointed this out.

BenjaminBlake-NOAA replied:
Good catch, I missed that somehow.

danielabdi-noaa replied:
Also both of these files are not available in current SRW app. So we would need to have those two fix files incorporated first. In my test, I pointed to the fix files from `RRFS_dev1` clone temporarily until a way is settled on how to incorporate the fix files.
```
  CUSTOM_POST_CONFIG_FP: '{{ [user.PARMdir, "upp/postxconfig-NT-fv3lam_rrfs.txt"]|path_join }}'
  CUSTOM_POST_PARAMS_FP: '{{ [user.PARMdir, "upp/params_grib2_tbl_new"]|path_join }}'
  TESTBED_FIELDS_FN: "testbed_fields_bgdawp.txt"
  TESTBED_FIELDS_FN2: "testbed_fields_bgrd3d.txt"
```
Moreover, these variables are first used by the `run_post` job so they should be defined there.
Another reason why order of tasks is important.

BenjaminBlake-NOAA replied:
I was thinking about the best way to handle this too.  I tested the prdgen task on WCOSS2 without these files and it completed successfully for me.  If the files aren't defined in the config.yaml file, then the testbed file generation part of the job should be skipped.  But once the run_post job is updated and the necessary files are added, this logic will already be in place for the testbed file generation and it can be turned on then.

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -2015,6 +2029,50 @@ task_run_post:
+#-----------------------------
+task_run_prdgen:
+  TN_RUN_PRDGEN: "run_prdgen"
+  NNODES_RUN_PRDGEN: 1
```
I believe the default should be 22 since that was what was hardcoded in the command
```
mpiexec -np 22 --cpu-bind core cfp $CMDFILE  #Daniel: Move this somewhere else
```

BenjaminBlake-NOAA replied:
Yeah setting the default as 22 sounds good to me.

danielabdi-noaa commented on tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_NA_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta.yaml:
```diff
@@ -37,3 +37,7 @@ task_run_fcst:
+task_run_prdgen:
+  RUN_TASK_RUN_PRDGEN: true
+  DO_PARALLEL_PRDGEN: true
+  PPN_RUN_PRDGEN: 22
```
I believe 22 should be assigned to `NNODES_RUN_PRDGEN` instead of `PPN_RUN_PRDGEN`

BenjaminBlake-NOAA replied:
Actually I intended for PPN_RUN_PRDGEN to be 22, so the task would run on 1 node but using 22 tasks per node, instead of 22 nodes with 1 task on each node.

danielabdi-noaa replied:
I am not sure about the parallel prdgen code. The script  generated on the fly that `mpiexec` calls launches 22 background jobs and waits, i.e. it is already parallel. However, we call the script with `mpiexec -np 22` or RUN_CMD_PRDGEN. Wouldn't that be  effectively calling 22x22 jobs?

BenjaminBlake-NOAA replied:
So the CFP module is actually what spreads out the tasks.  We define RUN_CMD_PRDGEN as "mpiexec -n $nprocs --cpu-bind core cfp".  In the ex-script, it gets called with eval $RUN_CMD_PRDGEN $CMDFILE, and cfp will execute each line of the script using 1 node.  For those 22 executions we just need 22 tasks on 1 node.

danielabdi-noaa replied:
Well this doesn't work on Hera since RUN_CMD_PRDGEN's defintion does not have "cfp". The prdgen task failed with out of memory error
```
slurmstepd: error: *** JOB 42907778 ON h33m43 CANCELLED AT 2023-03-15T10:16:19 DUE TO TIME LIMIT ***
slurmstepd: error: Detected 1786 oom-kill event(s) in StepId=42907778.batch. Some of your processes may have been killed by the cgroup out-of-memory handler.

```
So I think the RUN_CMD_PRDGEN defintion for all systems other than wcoss2 should be executed serially (mpirun -np 1) since there is no cfp there.

BenjaminBlake-NOAA replied:
Yes I agree.  The job will run more slowly as a serial job without CFP, but it should still work - the wall clock will need to be longer though.  Since the primary location this job will run is on WCOSS2 I think that is acceptable though.  If we will only use CFP on WCOSS2, then run_prdgen.local.lua files are not needed for the other supported platforms.

Commit:
Changes to address Daniel's suggestions

danielabdi-noaa commented on parm/FV3LAM_wflow.xml:
```diff
@@ -1188,6 +1189,46 @@ the <task> tag to be identical to the ones above for other output times.
+    <var name="fhr">{% for h in range(0, fcst_len_hrs+1) %}{{ " %03d" % h }}{% endfor %}
+</var>
+    <task name="&TN_RUN_PRDGEN;{{ uscore_ensmem_name }}_f#fhr#" cycledefs="forecast" max
+tries="{{ maxtries_run_prdgen }}">                 
```
There is a mistake here.

BenjaminBlake-NOAA replied:
Oh yeah the formatting got messed up somehow here.  I'll fix that

Commit:
Fixing typo in xml file

gspetro-NOAA commented:
@MichaelLueken @BenjaminBlake-NOAA I'm all set on the doc side! Didn't do an actual approval review because I can't speak to the code side. ;) But presuming that's all set, I agree this PR is ready to go!

MichaelLueken commented:
@gspetro-NOAA Thanks for approving the changes to the doc side!

christinaholtNOAA commented:
@MichaelLueken Will this one be ready for merge this morning, as well? It could be helpful to get it merged into my PR for the next round of testing. Thanks!

MichaelLueken commented:
@christinaholtNOAA This PR likely won't be merged this morning:

 1. I still need to submit the Jenkins tests (I can go ahead and do so now),
 2. following our discussion from two weeks ago, we still need an approval from an EMC code manager, and
 3. there appears to be an open discussion still related to the `POST_OUTPUT_DOMAIN_NAME` vs `TMMARK` naming convention.

As noted in 1, I will go ahead and submit the Jenkins tests now.  Is this update fine with not including an approval from an EMC code manager and should we apply some additional pressure to @hu5970 for an answer to the still open discussion on `scripts/exregional_run_prdgen.sh`?

BenjaminBlake-NOAA commented:
@MichaelLueken  It would be good if an EMC code manager could also submit an approval.  @RatkoVasic-NOAA or @chan-hoo do either of you have time to review/approve this PR?  As for the POST_OUTPUT_DOMAIN_NAME vs TMMARK issue, I am not sure what would be best.

RatkoVasic-NOAA commented:
@BenjaminBlake-NOAA did you get opinion from Ming Hu for POST_OUTPUT_DOMAIN_NAME vs TMMARK?

BenjaminBlake-NOAA commented:
@RatkoVasic-NOAA  Not yet.  I thought we wanted to move away from using $TMMARK, but I'd like for Ming or others who are more knowledgeable to comment on that.

hu5970 commented:
$TMMARK is for NAM type spin-up cycle. The current RRFS is using RAP/HRRR type spin-up cycle. I did not explain those two type of sin-up cycle here so you will feel confused. To be simple, in current RRFS cycles, the $TMMARK is always "tm00", it is means the current cycle minus 0. So for a file "RRFS_CONUS.t10z.bgdawpf003.tm00.grib2", I will see this is 3-h forecast from 10Z cycle. So "$TMMARK" should be removed or replaced with the "POST_OUTPUT_DOMAIN_NAME".

MichaelLueken commented:
@hu5970 Thank you very much for confirming that replacing `$TMMARK` with `$POST_OUTPUT_DOMAIN_NAME` is the correct move to make!  This has been done in this PR.

@RatkoVasic-NOAA With confirmation from Ming, I think we can continue moving forward with this update.  Would either you or @chan-hoo be able to review and approve this PR, so that the second issue is addressed?

@BenjaminBlake-NOAA Once the Jenkins tests are complete, I think this work will be ready to be merged.

RatkoVasic-NOAA reviewed (approved):
Looks good to me.

MichaelLueken commented:
The Jenkins tests have passed.  With @RatkoVasic-NOAA's approval, I will now merge this work to develop.
